README.md
# Image Writer

_flasher_ is a simple Disk Image USB burner tool..

_It is currently a work in progress. Maintenance is done by <a href="https://www.github.com/DAFT-8/">DAFT-8</a> team._

## Dependencies

* This application is developed based on Python3 and GTK+3. Dependencies:
  * ```gir1.2-glib-2.0 gir1.2-gtk-3.0 dh-python python3-all python3-setuptools gettext```

## Run Application from Source

* Install dependencies :
  * ```sudo apt install gir1.2-glib-2.0 gir1.2-gtk-3.0 dh-python python3-all python3-setuptools gettext```
* Clone the repository :
  * ```git clone https://github.com/DAFT-8/flasher.git ~/flasher```
* Run application :
  * ```python3 ~/flasher/src/main.py```

## Build the package

* `sudo apt install devscripts git-buildpackage`
* `cd ~/flasher/`
* `sudo mk-build-deps -ir`
* `gbp buildpackage --git-ignore-new -us -uc`

## Screenshot(s)

![flasher](screenshot.png)


debian/changelog
flasher (0.0.2) unstable; urgency=medium

  * Fix overloading
  * Add LICENSE
  * Update README

 -- Yurin Doctrine <myemail@mail.com>  Wed, 27 Mar 2022 14:51:12 +0300


src/main.py
#!/usr/bin/python3

import sys, os
import gi

gi.require_version('Gtk', '3.0')
from gi.repository import GLib, Gio, Gtk

from MainWindow import MainWindow


class Application(Gtk.Application):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, application_id="tr.org.daft-8.flasher",
                         flags=Gio.ApplicationFlags.HANDLES_OPEN | Gio.ApplicationFlags.NON_UNIQUE, **kwargs)
        self.window = None

    def do_activate(self):
        self.window = MainWindow(self)

    def do_open(self, files, filecount, hint):
        if filecount == 1:
            file = files[0]
            if os.path.exists(file.get_path()):
                fileFormat = file.get_basename().split(".")[-1]
                if fileFormat == "iso":
                    self.window = MainWindow(self, file.get_path())
                else:
                    print("Only .iso files.")
            else:
                print("File not exists : " + file.get_path())
        else:
            print("Only one file.")


if __name__ == "__main__":
    app = Application()
    app.run(sys.argv)


src/ISOCopier.py
#!/usr/bin/python3

import subprocess, sys, os, time, stat, signal


class IsoCopy:
    def errMsg(self, msg=""):
        sys.stderr.write("\x1b[31;1mError: \x1b[;0m{}".format(msg))
        exit(8)

    def __init__(self, iso_path, drive):
        # Define variables with fallback
        self.isoTmpFolder = "/run/flasher-iso-tmp/"
        self.usbMountFolder = "/run/flasher-usb-tmp/"
        self.isoPath = iso_path
        self.drive = drive
        self.isoName = ""

        # SIGTERM signal
        signal.signal(signal.SIGTERM, self.receiveSignal)

        # Check variables
        if not os.path.isfile(iso_path):
            self.errMsg("ISO file not found")
        if not stat.S_ISBLK(os.stat(drive).st_mode):
            self.errMsg("{} is not a valid block device".format(drive))

    def receiveSignal(self, number, frame):
        subprocess.run(["sync"])
        exit(15)

    def run(self):
        self.readIsoName()
        self.formatDrive()

        self.mountFolders()
        self.copyFiles()
        self.installGrub()

        # if "windows" in self.isoName.lower():
        #    self.windowsISOAddition()

        self.finishEvent()

    def readIsoName(self):
        with open(self.isoPath, "rb") as file:
            file.seek(32808,
                      0)  # Go to Volume Descriptor (https://wiki.osdev.org/ISO_9660#The_Primary_Volume_Descriptor)
            self.isoName = file.read(32).decode("utf-8").strip()  # Read 32 Bytes

    def formatDrive(self):
        # Unmount the drive before writing on it
        subprocess.run(["sh", "-c", ("ls {}* | xargs umount -lf ".format(self.drive))])

        # Format USB to FAT32
        subprocess.run(["dd", "if=/dev/zero", "of={}".format(self.drive), "bs=512", "count=1"])
        subprocess.run(["parted", self.drive, "mktable", "msdos"])
        subprocess.run(["parted", self.drive, "mkpart", "primary", "fat32", "1", "100%"])
        subprocess.run(["wipefs", "-a", (self.drive + "1"), "--force"])
        subprocess.run(["mkfs.vfat", (self.drive + "1")])
        subprocess.run(["parted", self.drive, "set", "1", "boot", "on"])
        subprocess.run(["sync"])

    def mountFolders(self):
        # Unmount first if already mounted
        subprocess.run(["umount", self.isoTmpFolder])
        subprocess.run(["umount", self.usbMountFolder])
        subprocess.run(["umount", (self.drive + "1")])

        subprocess.run(["mkdir", self.isoTmpFolder])
        subprocess.run(["mount", "-o", "ro", "-t", "auto", self.isoPath, self.isoTmpFolder])

        subprocess.run(["mkdir", self.usbMountFolder])
        subprocess.run(["mount", (self.drive + "1"), self.usbMountFolder])

    def copyFiles(self):
        folders = os.listdir(self.isoTmpFolder)
        fileCount = len(folders)
        i = 0
        for file in folders:
            subprocess.run(["rsync", "--archive", "--no-links", "--quiet", "--no-D", "--acls",
                            self.isoTmpFolder + file, self.usbMountFolder])

            subprocess.run(["sync"])
            print("PROGRESS:{}:{}".format(i, fileCount))
            sys.stdout.flush()
            i += 1

    def installGrub(self):
        # Install GRUB
        for target in ["i386-pc", "x86_64-efi", "i386-efi"]:
            subprocess.run(["rm", "-rf", "/{}/boot/grub/{}".format(self.usbMountFolder, target)])
        subprocess.run(["rm", "-rf", "/{}/EFI".format(self.usbMountFolder)])
        subprocess.run(["grub-install", "--target=i386-pc", "--force", "--removable",
                        "--boot-directory=/{}/boot".format(self.usbMountFolder),
                        " --locales=\"\"" ,self.drive])
        subprocess.run(["grub-install", "--target=x86_64-efi", "--force", "--removable",
                        "--efi-directory=/{}/".format(self.usbMountFolder),
                        "--boot-directory=/{}/boot".format(self.usbMountFolder),
                        "--locales=", self.drive])
        subprocess.run(["sync"])

    def windowsISOAddition(self):
        with open(self.usbMountFolder + "/boot/grub/grub.cfg", "a") as grubcfg:
            grubcfg.write("ntldr /bootmgr\nchainloader +1\nboot")

    def finishEvent(self):
        # Unmount the temp folder
        subprocess.run(["umount", self.isoTmpFolder])
        subprocess.run(["umount", (self.drive + "1")])
        subprocess.run(["rm", "-rf", self.isoTmpFolder])
        subprocess.run(["rm", "-rf", self.usbMountFolder])


if __name__ == "__main__":
    # Iso write action
    if len(sys.argv) < 2:
        sys.stderr.write("Usage: {} [iso path] [drive]\n".format(sys.argv[0]))
        exit(1)
    i = IsoCopy(sys.argv[1], sys.argv[2])
    i.run()


src/MainWindow.py
import os, sys, subprocess, requests
import gi

gi.require_version('Gtk', '3.0')
from gi.repository import GLib, Gio, Gtk

from USBDeviceManager import USBDeviceManager

import locale
from locale import gettext as tr

# Translation Constants:
APPNAME = "flasher"
TRANSLATIONS_PATH = "/usr/share/locale"
SYSTEM_LANGUAGE = os.environ.get("LANG")

# Translation functions:
locale.bindtextdomain(APPNAME, TRANSLATIONS_PATH)
locale.textdomain(APPNAME)
locale.setlocale(locale.LC_ALL, SYSTEM_LANGUAGE)


class MainWindow:
    def __init__(self, application, file=""):
        # Gtk Builder
        self.builder = Gtk.Builder()

        # Translate things on glade:
        self.builder.set_translation_domain(APPNAME)

        # Import UI file:
        self.builder.add_from_file(os.path.dirname(os.path.abspath(__file__)) + "/../ui/MainWindow.glade")
        self.builder.connect_signals(self)

        # Window
        self.window = self.builder.get_object("window")
        self.window.set_position(Gtk.WindowPosition.CENTER)
        self.window.set_application(application)
        self.window.connect("destroy", self.onDestroy)
        self.defineComponents()

        # Variables
        self.isGUILocked = False
        self.writeMode = "ImageWriter.py"  # ImageWriter.py for DD Mode, ISOCopier.py for ISO Mode

        # Get inserted USB devices
        self.imgFilepath = file
        if file:
            self.lbl_btn_selectISOFile.set_label(file.split('/')[-1])

        self.usbDevice = []
        self.usbManager = USBDeviceManager()
        self.usbManager.setUSBRefreshSignal(self.listUSBDevices)
        self.listUSBDevices()

        # Set version
        # If not getted from __version__ file then accept version in MainWindow.glade file
        try:
            version = open(os.path.dirname(os.path.abspath(__file__)) + "/__version__").readline()
            self.dialog_about.set_version(version)
        except:
            pass

        self.dialog_about.set_program_name(tr("flasher"))

        # Set application:
        self.application = application

        # Show Screen:
        self.window.show_all()

        # Debian based only signals
        if self.isdebian():
            self.installation_window()

    # Window methods:
    def onDestroy(self, action):
        self.window.get_application().quit()

    def defineComponents(self):
        self.stack_windows = self.builder.get_object("stack_windows")

        self.list_devices = self.builder.get_object("list_devices")
        self.cmb_devices = self.builder.get_object("cmb_devices")
        self.btn_selectISOFile = self.builder.get_object("btn_selectISOFile")
        self.lbl_btn_selectISOFile = self.builder.get_object("lbl_btn_selectISOFile")
        self.cmb_modes = self.builder.get_object("cmb_modes")
        self.stack_buttons = self.builder.get_object("stack_buttons")
        self.btn_start = self.builder.get_object("btn_start")
        self.pb_writingProgess = self.builder.get_object("pb_writingProgress")

        # Dialog:
        self.dialog_write = self.builder.get_object("dialog_write")
        self.dialog_write.set_position(Gtk.WindowPosition.CENTER)
        self.dlg_lbl_filename = self.builder.get_object("dlg_lbl_filename")
        self.dlg_lbl_disk = self.builder.get_object("dlg_lbl_disk")
        self.dialog_about = self.builder.get_object("dialog_about")

    # USB Methods
    def listUSBDevices(self):
        if self.isGUILocked == True:
            return

        deviceList = self.usbManager.getUSBDevices()
        self.list_devices.clear()
        for device in deviceList:
            self.list_devices.append(device)

        self.cmb_devices.set_active(0)
        self.stack_buttons.set_visible_child_name("start")

        if len(deviceList) == 0:
            self.btn_start.set_sensitive(False)
        elif self.imgFilepath and self.isGUILocked == False:
            self.btn_start.set_sensitive(True)

    def btn_selectISOFile_clicked(self, button):
        dialog = Gtk.FileChooserDialog(
            tr("Select ISO File..."),
            action=Gtk.FileChooserAction.OPEN,
            buttons=(Gtk.STOCK_CANCEL, Gtk.ResponseType.CANCEL, Gtk.STOCK_OPEN, Gtk.ResponseType.OK)
        )

        fileFilter = Gtk.FileFilter()
        fileFilter.set_name("*.iso, *.img")
        fileFilter.add_pattern("*.iso")
        fileFilter.add_pattern("*.img")
        dialog.add_filter(fileFilter)

        dialog.show()
        response = dialog.run()
        if response == Gtk.ResponseType.OK:
            filepath = dialog.get_filename()

            self.imgFilepath = filepath
            self.lbl_btn_selectISOFile.set_label(filepath.split('/')[-1])
            self.fileType = filepath.split(".")[-1]

            if self.imgFilepath and len(self.usbDevice) > 0:
                self.btn_start.set_sensitive(True)

        dialog.destroy()

    def cmb_devices_changed(self, combobox):
        tree_iter = combobox.get_active_iter()
        if tree_iter:
            model = combobox.get_model()
            deviceInfo = model[tree_iter][:3]
            self.usbDevice = deviceInfo
        else:
            self.btn_start.set_sensitive(False)

    def cmb_modes_changed(self, combobox):
        tree_iter = combobox.get_active_iter()
        if tree_iter:
            model = combobox.get_model()
            self.writeMode = model[tree_iter][0]  # 0:DD, 1:Iso
            if self.writeMode == 0:
                self.writeMode = "ImageWriter.py"
            elif self.writeMode == 1:
                self.writeMode = "ISOCopier.py"
            elif self.writeMode == 2:
                self.writeMode = "WinUSB.py"
            if self.writeMode != 0:
                if not os.path.isdir("/usr/lib/grub/i386-pc"):
                    combobox.set_active(0)
                    self.writeMode = 0
                    if not self.isdebian():
                        self.show_message(tr("Target {} does not exists").format("i386-pc"),
                                          tr("Please install {}").format("grub-pc-bin"))
                    else:
                        self.builder.get_object("mode_installer").set_current_page(1)
                    self.builder.get_object("mode_label").set_text(tr("{} not found").format("grub-i386-pc"))
                elif not os.path.isdir("/usr/lib/grub/x86_64-efi"):
                    combobox.set_active(0)
                    self.writeMode = 0
                    if not self.isdebian():
                        self.show_message(tr("Target {} does not exists").format("x86_64-efi"),
                                          tr("Please install {}").format("grub-efi-amd64-bin"))
                    else:
                        self.builder.get_object("mode_installer").set_current_page(1)
                    self.builder.get_object("mode_label").set_text(tr("{} not found").format("grub-x86_64-amd64-efi"))

    # Buttons:
    def btn_start_clicked(self, button):
        self.prepareWriting()

    def btn_cancel_clicked(self, button):
        self.cancelWriting()

    def btn_exit_clicked(self, button):
        self.window.get_application().quit()

    def btn_write_new_file_clicked(self, button):
        self.stack_windows.set_visible_child_name("main")

    def btn_information_clicked(self, button):
        self.dialog_about.run()
        self.dialog_about.hide()

    def show_message(self, msg1="", msg2=""):
        dialog = Gtk.MessageDialog(
            self.window,
            0,
            Gtk.MessageType.ERROR,
            Gtk.ButtonsType.OK,
            tr(msg1),
        )
        dialog.format_secondary_text(
            tr(msg2)
        )
        dialog.run()
        dialog.destroy()

    def isdebian(self):
        return os.path.exists("/var/lib/dpkg/status")

    def installation_window(self):
        yes = self.builder.get_object("but_inst")
        no = self.builder.get_object("but_canc")
        nm = self.builder.get_object("mode_installer")
        nm.set_current_page(0)

        def yes_event(widget):
            nm.set_current_page(2)
            apt_install_grub()

        def no_event(widget):
            nm.set_current_page(0)

        def apt_install_grub():
            params = "pkexec apt install --reinstall grub-pc-bin grub-efi-amd64-bin -yq".split(" ")

            def onProcessExit(pid, status):
                if status == 0:
                    self.builder.get_object("mod_message").set_text(tr("Installation done."))
                else:
                    self.builder.get_object("mod_message").set_text(tr("Installation failed."))
                nm.set_current_page(3)
                yes.set_sensitive(True)

            try:
                writerProcessPID, _, stdout, _ = GLib.spawn_async(params,
                                                                  flags=GLib.SPAWN_SEARCH_PATH | GLib.SPAWN_LEAVE_DESCRIPTORS_OPEN | GLib.SPAWN_DO_NOT_REAP_CHILD,
                                                                  standard_input=False, standard_output=True,
                                                                  standard_error=False)
            except:
                nm.set_current_page(3)
                self.builder.get_object("mod_message").set_text(tr("Installation failed."))
            GLib.child_watch_add(GLib.PRIORITY_DEFAULT, writerProcessPID, onProcessExit)

        yes.connect("clicked", yes_event)
        no.connect("clicked", no_event)
        self.builder.get_object("go_back").connect("clicked", no_event)

    def startWriting(self):
        self.lockGUI()
        self.startProcess([
            "pkexec",
            os.path.dirname(os.path.abspath(__file__)) + "/" + self.writeMode,
            self.imgFilepath,
            '/dev/' + self.usbDevice[0],
        ])

    def prepareWriting(self):
        # Ask if it is ok?
        self.dlg_lbl_filename.set_markup("- <b>{}</b>".format(self.imgFilepath.split('/')[-1]))
        self.dlg_lbl_disk.set_markup(
            "- <b>{} [ {} ]</b> <i>( /dev/{} )</i>".format(self.usbDevice[1], self.usbDevice[2], self.usbDevice[0]))

        response = self.dialog_write.run()
        self.dialog_write.hide()
        if response == Gtk.ResponseType.YES:
            self.startWriting()

    def cancelWriting(self):
        subprocess.call(["pkexec", "kill", "-SIGTERM", str(self.writerProcessPID)])

    def startProcess(self, params):
        self.writerProcessPID, _, stdout, _ = GLib.spawn_async(params,
                                                               flags=GLib.SPAWN_SEARCH_PATH | GLib.SPAWN_LEAVE_DESCRIPTORS_OPEN | GLib.SPAWN_DO_NOT_REAP_CHILD,
                                                               standard_input=False, standard_output=True,
                                                               standard_error=False)

        GLib.io_add_watch(GLib.IOChannel(stdout), GLib.IO_IN | GLib.IO_HUP, self.onProcessStdout)

        GLib.child_watch_add(GLib.PRIORITY_DEFAULT, self.writerProcessPID, self.onProcessExit)

    def onProcessStdout(self, source, condition):
        if condition == GLib.IO_HUP:
            return False

        line = source.readline().strip()

        if self.writeMode == "ImageWriter.py":
            written, total = line.split()
            written = int(written)
            total = int(total)
            percent = 0
            if total > 0:
                percent = written / total

            self.pb_writingProgess.set_text(
                "{}MB / {}MB (%{})".format(round(written / 1000 / 1000), round(total / 1000 / 1000),
                                           int(percent * 100)))
            self.pb_writingProgess.set_fraction(percent)
        else:
            if line[0:9] == "PROGRESS:":
                _, copied, total = line.split(":")
                copied = int(copied)
                total = int(total)

                percent = 0
                if total > 0:
                    percent = copied / total

                self.pb_writingProgess.set_text("%{}".format(int(percent * 100)))
                self.pb_writingProgess.set_fraction(percent)
        return True

    def onProcessExit(self, pid, status):
        self.unlockGUI()
        self.listUSBDevices()
        self.pb_writingProgess.set_fraction(0)

        if status == 0:
            self.pb_writingProgess.set_text("0%")
            self.sendNotification(tr("Writing process is finished."), tr("You can eject the USB disk."))
            self.stack_windows.set_visible_child_name("finished")
        elif status != 15 and status != 32256:  # these are cancelling or auth error.
            self.pb_writingProgess.set_text(tr("Error!"))
            self.pb_writingProgess.set_fraction(0)
            dialog = Gtk.MessageDialog(
                self.window,
                0,
                Gtk.MessageType.ERROR,
                Gtk.ButtonsType.OK,
                tr("An error occured while writing the file to the disk."),
            )
            dialog.format_secondary_text(
                tr("Please make sure the USB device is connected properly and try again.")
            )
            dialog.run()
            dialog.destroy()

    def lockGUI(self, disableStart=False):
        self.btn_selectISOFile.set_sensitive(False)
        self.cmb_devices.set_sensitive(False)
        self.cb_checkIntegrity.set_sensitive(False)
        self.cmb_modes.set_sensitive(False)

        self.stack_buttons.set_visible_child_name("cancel")
        self.isGUILocked = True

    def unlockGUI(self):
        self.btn_selectISOFile.set_sensitive(True)
        self.cmb_devices.set_sensitive(True)
        self.cb_checkIntegrity.set_sensitive(True)
        self.cmb_modes.set_sensitive(True)

        self.stack_buttons.set_visible_child_name("start")
        self.isGUILocked = False

    def sendNotification(self, title, body):
        notification = Gio.Notification.new(title)
        notification.set_body(body)
        notification.set_icon(Gio.ThemedIcon(name="flasher"))
        notification.set_default_action("app.notification-response::focus")
        self.application.send_notification(self.application.get_application_id(), notification)
